<a href="https://colab.research.google.com/github/vsnupoudel/RandomAssignments/blob/main/Excel_File_manipulation_from_Python.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [23]:
import pandas as pd
import openpyxl as xl
import requests
import os
# open('facebook.ico', 'wb').write(r.content)

## Download the file “Financial_Report.xlsx” from the link (do it using Python code to download the file mentioned above)


In [24]:
file_name = 'Financial_Report.xlsx'
if os.path.exists('./'+file_name):
    os.remove('./'+file_name)

In [25]:
url = 'https://www.sec.gov/Archives/edgar/data/789019/000156459021002316/Financial_Report.xlsx'
with open(file_name, "wb+") as file:
        # get request
        response = requests.get(url)
        # write to file
        file.write(response.content)

# The url returns HTML, instead of excel when this cell is re-run. 
# I believe, It might be a collab related issue, so I will look into that later.

## Uploading the file manually if it shows error in this step

In [26]:
balsheet = pd.read_excel('./Financial_Report.xlsx'
                            , sheet_name='BALANCE SHEETS'
                              , index_col = 0
                         )

## Go to the “Balance Sheets” sheet from the above file and extract only rows corresponding to Current Liabilities.

In [30]:
listOfindex = list(balsheet.index )
# listOfindex

In [31]:
import regex as re
start = [listOfindex.index(i) for i in listOfindex if ('current' in i.lower().strip()[0:7]
                     and 'liabilities'  in i.lower().strip()[-15:] ) ] 
start = start[0]
start

15

In [32]:
end = [listOfindex.index(i) for i in listOfindex if ('stockholder' in i.lower().strip()[0:11]
                     and 'equity'  in i.lower().strip()[-7:] ) ] 
end = end[0]
end

31

In [33]:
cur_liab = balsheet.iloc[start:end , :]

In [35]:
cur_liab.sample(2)

Dec. 31, 2020 Jun. 30, 2020
BALANCE SHEETS - USD ($) $ in Millions                            
Other long-term liabilities                    12544         10632
Operating lease liabilities                     8875          7671

## What proportion of “Total Current Liabilities” is “Accounts Payable”?

In [36]:
cur_liab = cur_liab[['Dec. 31, 2020'  , 'Jun. 30, 2020' ]].replace(' ', 0)

In [37]:
cur_liab.astype({  'Dec. 31, 2020' : float 
                 , 'Jun. 30, 2020' : float  })

Dec. 31, 2020  Jun. 30, 2020
BALANCE SHEETS - USD ($) $ in Millions                              
Current liabilities:                              NaN            NaN
Accounts payable                              12770.0        12530.0
Current portion of long-term debt              5387.0         3749.0
Accrued compensation                           6838.0         7874.0
Short-term income taxes                        1562.0         2130.0
Short-term unearned revenue                   30402.0        36000.0
Other current liabilities                     10527.0        10027.0
Total current liabilities                     67486.0        72310.0
Long-term debt                                55136.0        59578.0
Long-term income taxes                        26701.0        29432.0
Long-term unearned revenue                     2985.0         3180.0
Deferred income taxes                           174.0          204.0
Operating lease liabilities                    8875.0         7671.0
Other long-term liabilities                   12544.0        10632.0
Total liabilities                            173901.0       183007.0
Commitments and contingencies                     0.0            0.0

In [38]:
cur_liab.loc[ 'Accounts payable' , ].sum() /  cur_liab.sum().sum()

0.029638758592897007

### Answer : 0.02964

## By how much percent is “Total Liabilities and stockholders equity” for  Dec, 31st 2020 more or less than Jun, 30th 2020?
- Corrected the wording  of question slightly

In [42]:
( balsheet.loc['Total liabilities and stockholders’ equity', 'Dec. 31, 2020'] - balsheet.loc[
                        'Total liabilities and stockholders’ equity', 'Jun. 30, 2020']) / balsheet.loc[
            'Total liabilities and stockholders’ equity', 'Dec. 31, 2020'] 

0.00929186517917912

- It is 0.93% or almost 1% more

## Combine all the sheets with the same column names (Dec, 31st 2020 and June 30th 2020) and create a master excel sheet with these two columns. The final master excel sheet should have four columns: the two dates mentioned above, variable name (like Cash and Cash Equivalents) and variable type (like Current Assets).

Answer : Should be done in VBA so that it can be easily automated later. VBA code could be called from python script